### PR TITLE
Realization from Galacticus output

### PR DESCRIPTION
**Pull request by cgannonucm**
Implements the ability to create a realization from a Galacticus output file.

- TNFWFromParams (class)
- DMFromGalacticus (method)
- pyHalo/Halo/galacticus_util (Utilities for working with Galacticus output)

### CODE REVIEW

File: pyHalo/Halos/HaloModels/TNFWFromParams.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+        self._lens_cosmo = lens_cosmo_instance
+
+        self._kpc_per_arcsec_at_z = self._lens_cosmo.cosmo.kpc_proper_per_asec(z)
+
+        x = x_kpc / self._kpc_per_arcsec_at_z
+
+        y = y_kpc / self._kpc_per_arcsec_at_z
+
+        self._params_physical = args
```
Comment (dangilman) [inline pyHalo/Halos/HaloModels/TNFWFromParams.py:34]: In general args might have other parameters besides the physical parameters describing the NFW profile. Can you extract only the parameters you need from args to define self._params_physical?

File: pyHalo/Halos/HaloModels/TNFWFromParams.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        r_s = _params[self.KEY_RS]
+        rho_s = _params[self.KEY_RHO_S]
+
+        n = 1
+
+
+        x = r / r_s
+        tau = r_t / r_s
+
+        n = 0
```
Comment (dangilman) [inline pyHalo/Halos/HaloModels/TNFWFromParams.py:60]: why n=0?

File: pyHalo/Halos/HaloModels/TNFWFromParams.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        return self._params_physical
+
+    @property
+    def lenstronomy_params(self):
+        """
+        See documentation in base class (Halos/halo_base.py)
+        """
+        if not hasattr(self, '_kwargs_lenstronomy'):
+            [concentration, rt] = self.profile_args
+            Rs_angle, theta_Rs = self._lens_cosmo.nfw_physical2angle(self.mass, concentration, self.z)
```
Comment (dangilman) [inline pyHalo/Halos/HaloModels/TNFWFromParams.py:101]: This method will have to change to compute Rs_angle and theta_Rs directly from self._params_physical

**Review by dangilman — CHANGES_REQUESTED** [review]
Hi @cgannonucm, looks like a reasonable implementation to me! I left a few comments just on the new Halo class for now. 

First, since some of the methods are identical to the other TNFW halo class I think you should have this class be a subclass of TNFW, i.e. TNFWFromParams(TNFW) instead of TNFWFromParams(Halo). That way you don't have to copy the same method into different classes. Note that you will have override some methods in the new class, in particular the params_physical calculation, calculating the concentration, and the lenstronomy_params method. 

On the last point, you'll need to compute theta_Rs and Rs_angle directly from params physical rather than using the mass and concentration. You can use the methods in lens_cosmo for this

File: pyHalo/Halos/HaloModels/TNFWFromParams.py (hunk 2 above)
Comment (cgannonucm) [reply pyHalo/Halos/HaloModels/TNFWFromParams.py:60]: Good catch, this was left in from when I was running some checks. Will remove.

Comment (cgannonucm) [reply pyHalo/Halos/HaloModels/TNFWFromParams.py:60]: n should be 1

**Comment by dangilman** [comment]
Hi @cgannonucm, looks good to me. For some reason your pull request doesn't trigger tests and the coveralls run, I'm not sure why. In the meantime, it looks like there are some merge conflicts with the main branch that need to be resolved. In particular, the preset model functions are now all distributed in their files in the PresetModels class folder -- you can put this one in the external.py file. Maybe if you fix the merge conflicts it will trigger the tests and coverage runs.....

**Comment by cgannonucm** [comment]
Hi @dangilman  This branch and the main branch should be merged now, and all conflicts resolved.

**Comment by dangilman** [comment]
@cgannonucm thanks! It looks like there is a problem with one of the test functions for a different method that is causing the tests to fail, would you mind fixing it so that the coverage and tests run? You just have to change the function names in the import statement of test_realization_extensions.py from _xi_l, _xi_l_to_Pk_l to "xi_l" and "xi_l_to_Pk_l". Then it should work

**Comment by cgannonucm** [comment]
@dangilman Done. The tests should run now.

**Review by dangilman — APPROVED** [review]
Looks great, thanks @cgannonucm !

**Pull request merged**